lib/Lib.IntTypes.fst

```diff
@@ -374,25 +668,88 @@ let eq_mask #t a b =
   | U32  -> UInt32.eq_mask a b
   | U64  -> UInt64.eq_mask a b
   | U128 -> UInt128.eq_mask a b
-
-let eq_mask_lemma #t a b =
+  | S8   -> Int.Cast.uint8_to_int8 (UInt8.eq_mask (to_u8 a) (to_u8 b))
+  | S16  -> Int.Cast.uint16_to_int16 (UInt16.eq_mask (to_u16 a) (to_u16 b))
+  | S32  -> Int.Cast.uint32_to_int32 (UInt32.eq_mask (to_u32 a) (to_u32 b))
+  | S64  -> Int.Cast.uint64_to_int64 (UInt64.eq_mask (to_u64 a) (to_u64 b))
+
+val eq_mask_lemma_unsigned: #t:inttype{unsigned t} -> a:int_t t SEC -> b:int_t t SEC -> Lemma
+  (if v a = v b then v (eq_mask a b) == ones_v t
+                else v (eq_mask a b) == 0)
+let eq_mask_lemma_unsigned #t a b =
   match t with
   | U1 ->
-    begin
     assert_norm (
-      logxor (u1 0) (u1 0) == u1 0 /\ logxor (u1 1) (u1 1) == u1 0 /\
-      logxor (u1 0) (u1 1) == u1 1 /\ logxor (u1 1) (u1 0) == u1 1 /\
+      logxor (u1 0) (u1 0) == u1 0 /\ logxor (u1 0) (u1 1) == u1 1 /\
+      logxor (u1 1) (u1 0) == u1 1 /\ logxor (u1 1) (u1 1) == u1 0 /\
       lognot (u1 1) == u1 0 /\ lognot (u1 0) == u1 1)
+  | U8 | U16 | U32 | U64 | U128 -> ()
+
+val eq_mask_lemma_signed: #t:inttype{signed t /\ ~(S128? t)} -> a:int_t t SEC -> b:int_t t SEC -> Lemma
+  (if v a = v b then v (eq_mask a b) == ones_v t
+                else v (eq_mask a b) == 0)
+let eq_mask_lemma_signed #t a b =
```

Comment:
This does not typecheck on my computer (Z3 4.8.5 says: unknown because unknown).
Someone has the same issue ?

Reply:
It works for me, but it's close to the z3rlimit. I doubled it to be safe.